activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -76,19 +76,21 @@ def sum(identity = nil, &block)
       _original_sum_with_required_identity(identity, &block)
     elsif block_given?
       map(&block).sum
-    # we check `first(1) == []` to check if we have an
-    # empty Enumerable; checking `empty?` would return
-    # true for `[nil]`, which we want to deprecate to
-    # keep consistent with Ruby
-    elsif first.is_a?(Numeric) || first(1) == [] || first.respond_to?(:coerce)
-      identity ||= 0
-      _original_sum_with_required_identity(identity, &block)
     else
-      ActiveSupport::Deprecation.warn(<<-MSG.squish)
-        Rails 7.0 has deprecated Enumerable.sum in favor of Ruby's native implementation available since 2.4.
-        Sum of non-numeric elements requires an initial argument.
-      MSG
-      inject(:+) || 0
+      enum = enum_for(:each)
+      first_value =
+        begin
+          first_value = enum.next
+        rescue StopIteration
+          return 0
+        end
+      unless first_value.is_a?(Numeric) || first_value.respond_to?(:coerce)
+        ActiveSupport::Deprecation.warn(<<-MSG.squish)
+          Rails 7.0 has deprecated Enumerable.sum in favor of Ruby's native implementation available since 2.4.
+          Sum of non-numeric elements requires an initial argument.
+        MSG
+      end
+      Enumerator.produce(first_value) { enum.next }.inject(:+) || 0
```

Comment:
This is just my opinion, but I feel like @matthewd's second suggestion is more straightforward (and more efficient).  If we want a less `for`-loop-like implementation, we could also use `reduce` (aka `inject`) instead of `each`:

```ruby
reduce(nil) do |sum, value|
  if sum.nil?
    unless value.is_a?(Numeric) || value.respond_to?(:coerce)
      # warn
    end
    value
  else
    sum + value
  end
end || 0
```

Reply:
That assumes the enumerable doesn't start with nil values / `sum + value` never returns nil.

Reply:
(I agree the Enumerator version is probably overcomplicated / needlessly inefficient)

Reply:
> That assumes the enumerable doesn't start with nil values / `sum + value` never returns nil.

When we have `nil`s at the beginning (`[nil, nil, 2, 3, 4]`), the branch with `if sum.nil?` will be executed for all of them. When we have `nil` in the middle somewhere (`[1, 2, nil, 3]`), we will get an error about not able to coerce, like in native `sum`. Am I missing something?

Reply:
> the branch with if sum.nil? will be executed for all of them

That's a behaviour change.

(More obscurely, a redefined `NilClass#+` would also deviate.)

Reply:
> That assumes the enumerable doesn't start with nil values / `sum + value` never returns nil.

It does, but we want `[nil].sum == 0`:

https://github.com/rails/rails/blob/f61bcfb315c31f042a343249995c7e6e21e64334/activesupport/test/core_ext/enumerable_test.rb#L140-L142

I don't think `+` would purposefully return `nil` because `nil + anything` would raise (unless you override `NilClass#+`, which seems... unwise).  But I suppose we could be masking an error here.  Alternatively:

```ruby
first = true

reduce(nil) do |sum, value|
  if first
    unless value.is_a?(Numeric) || value.respond_to?(:coerce)
      # warn
    end

    first = false
    value
  else
    sum + value
  end
end || 0
```

Reply:
> we want [nil].sum == 0

Ah, fair.

> unless you override NilClass#+, which seems... unwise

Agree, but even while "unwise" might be a reasonable line in main, I think we should aim for unreasonable accomodation in stable branches.

(And I _can_ imagine a risky but loosely-plausible NilClass#+ that exclusively accepts some specific CustomClass as its argument)

Reply:
Updated with the `if first` variant.